server/apis/v1/handler.go

```diff
@@ -46,282 +68,831 @@ type handler struct {
 
 // NewHandler is used to provide a new instance of the handler type
 func NewHandler() (*handler, error) {
-	restConfig, err := sharedutil.K8sRestConfig()
+	var (
+		k8sRestConfig *rest.Config
+		err           error
+	)
+	k8sRestConfig, err = util.K8sRestConfig()
 	if err != nil {
-		return nil, fmt.Errorf("failed to get kubeconfig, %w", err)
+		return nil, fmt.Errorf("failed to get kubeRestConfig, %w", err)
 	}
-	kubeClient, err := kubernetes.NewForConfig(restConfig)
+	kubeClient, err := kubernetes.NewForConfig(k8sRestConfig)
 	if err != nil {
 		return nil, fmt.Errorf("failed to get kubeclient, %w", err)
 	}
-	metricsClient := metricsversiond.NewForConfigOrDie(restConfig)
-	numaflowClient := dfv1versiond.NewForConfigOrDie(restConfig).NumaflowV1alpha1()
+	metricsClient := metricsversiond.NewForConfigOrDie(k8sRestConfig)
+	numaflowClient := dfv1versiond.NewForConfigOrDie(k8sRestConfig).NumaflowV1alpha1()
 	return &handler{
 		kubeClient:     kubeClient,
 		metricsClient:  metricsClient,
 		numaflowClient: numaflowClient,
 	}, nil
 }
 
+// ListNamespaces is used to provide all the namespaces that have numaflow pipelines running
+func (h *handler) ListNamespaces(c *gin.Context) {
+	namespaces, err := getAllNamespaces(h)
+	if err != nil {
+		h.respondWithError(c, fmt.Sprintf("Failed to fetch all namespaces, %s", err.Error()))
+		return
+	}
+	c.JSON(http.StatusOK, NewNumaflowAPIResponse(nil, namespaces))
+}
+
+// GetClusterSummary summarizes information of all the namespaces in a cluster and wrapped the result in a list.
+func (h *handler) GetClusterSummary(c *gin.Context) {
+	namespaces, err := getAllNamespaces(h)
+	if err != nil {
+		h.respondWithError(c, fmt.Sprintf("Failed to fetch cluster summary, %s", err.Error()))
+		return
+	}
+	var clusterSummary ClusterSummaryResponse
+	// Loop over the namespaces to get status
+	for _, ns := range namespaces {
```

Comment:
This is too low efficient, the way to do it should be get all the pipelines of the cluster with 1 call, and then do the loop in memory. Please add a `TODO` to change it. @jy4096 @kohlisid